### PR TITLE
Avoid using db.stateCache.values()

### LINKED ISSUE #1118: Profiling: archiveStates is slow
While syncing Altona, CPU profiler shows significant time spent in `archiveStates`, specifically, because we call `stateCache.values`, which iterates thru all states/contexts and copies them. Even though this is relatively lightweight, it adds up.

**Is your feature request related to a problem? Please describe.**
Slow `stateCache.values` is a bottleneck.

**Describe the solution you'd like**
avoid calling `stateCache.values` alltogether. Ideally, we should be able to use the information from the `forkChoice` (with some minor refactoring) to figure out which stateRoots to remove. Currently, it seems we immediately delete the old `forkChoice` data, and only delete the old `stateCache` afterwards as a "task".

**Describe alternatives you've considered**
Make copying/cloning cheaper (we should try to do this too)

### PR DESCRIPTION
**Pull request by twoeths**
resolves #1118 

+ We know that using `db.stateCache.values()` is expensive so just comment out that for now. The processing time is reduce from ~2500ms to ~200ms in my environment.

### CODE REVIEW

**Comment by wemeetagain** [comment]
I'd like to see about using the technique described in the issue.
Fork choice data mirrors the data in `db.stateCache` and `db.block`. The work of knowing "which to delete" is already done in `forkChoice.prune`. This would make it trivial to prune the cache:
```typescript
// archive state
forkChoice.on("prune", async (summaries: BlockSummary[]) => {
  await db.stateArchive.set(...);
  await db.stateCache.batchDelete(summaries.map(s => s.stateRoot));
});
// archive blocks
forkChoice.on("prune", async (summaries: BlockSummary[]) => {
  await db.blockArchive.set(...);
  await db.block.batchDelete(summaries.map(s => s.blockRoot));
});
```

**Comment by mpetrunic** [comment]
I think fork choice prune isnt the same as cache prune aince fork choice keeps last finalized stuff.

**Comment by twoeths** [comment]
just double check:
+ For fork choice, we keep finalized stuff, remove nodes < `finalized.slot`
+ For ArchiveState, we keep finalized state in `stateCache`, remove states < `finalized.slot`
+ For ArchiveBlock, we only use `finalized`, not `pruned` so it's just the same to the current implementation

so we're safe

File: packages/lodestar/src/chain/chain.ts [hunk 1]
```diff
@@ -232,7 +232,7 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
 
   public async waitForBlockProcessed(blockRoot: Uint8Array): Promise<void> {
     await new Promise((resolve) => {
-      this.once("processedBlock", (signedBlock) => {
+      this.on("processedBlock", (signedBlock) => {
         const root = this.config.types.BeaconBlock.hashTreeRoot(signedBlock.message);
         if (this.config.types.Root.equals(root, blockRoot)) {
           resolve();
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/chain.ts:239]: you should probably unsubscribe processedBlock event at this point

File: packages/lodestar/src/db/api/beacon/stateContextCache.ts [hunk 2]
```diff
@@ -42,9 +46,13 @@ export class StateContextCache {
...
-  public async values(): Promise<ITreeStateContext[]> {
-    return Object.values(this.cache).map(item => this.clone(item));
-  }
+  /**
+   * Should only use this with care as this is expensive.
+   * @param epoch
+   */
+  // public async values(): Promise<ITreeStateContext[]> {
+  //   return Object.values(this.cache).map(item => this.clone(item));
+  // }
```
Comment (mpetrunic) [inline packages/lodestar/src/db/api/beacon/stateContextCache.ts]: We could maybe call it `valuesUnsafe` or something though

**Comment by twoeths** [comment]
I got this error after syncing up to block 4968
```
(node:28904) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'state' of null
    at BeaconChain.getHeadState (/Users/tuyennguyen/Projects/workshop/lodestar/packages/lodestar/lib/chain/chain.js:100:76)
    at async LocalClock.<anonymous> (/Users/tuyennguyen/Projects/workshop/lodestar/packages/lodestar/lib/sync/utils/attestation-collector.js:29:25)
```
need to investigate that ...

**Comment by mpetrunic** [comment]
> I got this error after syncing up to block 4968
> 
> ```
> (node:28904) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'state' of null
>     at BeaconChain.getHeadState (/Users/tuyennguyen/Projects/workshop/lodestar/packages/lodestar/lib/chain/chain.js:100:76)
>     at async LocalClock.<anonymous> (/Users/tuyennguyen/Projects/workshop/lodestar/packages/lodestar/lib/sync/utils/attestation-collector.js:29:25)
> ```
> 
> need to investigate that ...

That's slot in which we have state transition split (we produce state with different root).
@wemeetagain already open PRs with fix for this in ssz.

File: packages/lodestar/src/tasks/tasks/archiveBlocks.ts [hunk 3]
```diff
@@ -23,31 +25,31 @@ export class ArchiveBlocksTask implements ITask {
...
-    this.finalizedCheckpoint = finalizedCheckpoint;
+    this.finalized = finalized;
+    this.pruned = pruned;
   }
 
   /**
    * Only archive blocks on the same chain to the finalized checkpoint.
    */
   public async run(): Promise<void> {
+    this.logger.profile("Archieve Blocks");
```
Comment (wemeetagain) [inline packages/lodestar/src/tasks/tasks/archiveBlocks.ts]: typo, the other logger messages here and in archiveState are typo'd too

**Review by wemeetagain — CHANGES_REQUESTED** [review]
typos, this looks great tho!

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**